api/firstbook2.py

```diff
@@ -0,0 +1,196 @@
+
+from authenticator import (
+    BasicAuthenticationProvider,
+    PatronData,
+)
+from config import (
+    Configuration,
+    CannotLoadConfiguration,
+)
+from circulation_exceptions import RemoteInitiatedServerError
+import urlparse
+import urllib
+from core.model import (
+    get_one_or_create,
+    ExternalIntegration,
+    Patron,
+)
+
+class FirstBookAuthenticationAPI(BasicAuthenticationProvider):
+
+    NAME = 'First Book - JWT'
+
+    DESCRIPTION = _("""
+        An authentication service for Open eBooks that authenticates
+        using access codes and PINs. (This is the new version.)""")
+
+    DISPLAY_NAME = NAME
+    DEFAULT_IDENTIFIER_LABEL = _("Access Code")
+    LOGIN_BUTTON_IMAGE = "FirstBookLoginButton280.png"
+
+    # The algorithm used to sign JWTs.
+    ALGORITHM = 'HS256'
+
+    # If FirstBook sends this message it means they accepted the
+    # patron's credentials.
+    SUCCESS_MESSAGE = 'Valid Code Pin Pair'
+
+    # Server-side validation happens before the identifier
+    # is converted to uppercase, which means lowercase characters
+    # are valid.
+    DEFAULT_IDENTIFIER_REGULAR_EXPRESSION = '^[A-Za-z0-9@]+$'
+    DEFAULT_PASSWORD_REGULAR_EXPRESSION = '^[0-9]+$'
+
+    SETTINGS = [
+        {
+            "key": ExternalIntegration.URL, "label": _("URL"),
+            "default": "https://ebooksuat.firstbook.org/api/",
+            "required": True
+        },
+        { "key": ExternalIntegration.PASSWORD, "label": _("Key"), "required": True },
+    ] + BasicAuthenticationProvider.SETTINGS
+
+    log = logging.getLogger("First Book JWT authentication API")
+
+    def __init__(self, library_id, integration, analytics=None, root=None,
+                 secret=None):
+        super(FirstBookAuthenticationAPI, self).__init__(library_id, integration, analytics)
+        root = root or integration.url
+        secret = secret or integration.password
+        if not (root and secret):
+            raise CannotLoadConfiguration(
+                "First Book server not configured."
+            )
+        self.root = root
+        self.secret = secret
+
+    # Begin implementation of BasicAuthenticationProvider abstract
+    # methods.
+
+    def remote_authenticate(self, username, password):
+        # All FirstBook credentials are in upper-case.
+        username = username.upper()
+
+        # If they fail a PIN test, there is no authenticated patron.
+        if not self.remote_pin_test(username, password):
+            return None
+
+        # FirstBook keeps track of absolutely no information
+        # about the patron other than the permanent ID,
+        # which is also the authorization identifier.
+        return PatronData(
+            permanent_id=username,
+            authorization_identifier=username,
+        )
+
+    # End implementation of BasicAuthenticationProvider abstract methods.
+
+    def remote_pin_test(self, barcode, pin):
+        jwt = self.jwt(barcode, pin)
+        url = self.root + jwt
```

Comment:
This is new -- previously we stuck the barcode and PIN on to the end of `url`; now we call `jwt` and put _that_ on the end.